File: algorithm_data_structure/leetcode/142.linked-list-cycle-ii.py
#
# @lc app=leetcode id=142 lang=python3
#
# [142] Linked List Cycle II
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def detectCycle(self, head: ListNode) -> ListNode:
        slow = head
        fast = head

        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
            if fast == slow:
                break

        # There's no cycle
        if not fast or not fast.next:
            return None

        # There's cycle
        slow = head
        while slow != fast:
            slow = slow.next
            fast = fast.next
        return fast


# @lc code=end

File: algorithm_data_structure/leetcode/142.linked-list-cycle-ii.cpp
/*
 * @lc app=leetcode id=142 lang=cpp
 *
 * [142] Linked List Cycle II
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
   public:
	ListNode *detectCycle(ListNode *head) {
		// Linked-list -> Slow/fast pointer
		ListNode *slow = head;
		ListNode *fast = head;

		do {
			if (!fast || !fast->next) {
				return nullptr;
			}
			fast = fast->next->next;
			slow = slow->next;
		} while (fast != slow);

		// Exist the loop! So there is a cycle
		fast = head;
		// Let the fast pointer starts from the head
		// When slow and fast meets again, this is the postion of the loop.
		while (fast != slow) {
			fast = fast->next;
			slow = slow->next;
		}
		return fast;
	}
};
// @lc code=end
